experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-translator.ts

```diff
@@ -1,11 +1,184 @@
- * 2. Receives requests from Azure Web PubSub requests and translate them into Engine.IO behaviours.
- * 3. Translates Engine.IO behaviours to Azure Web PubSub service behaviours like REST API calls.
- * 4. Makes the Engine.IO `sid` same as its corresponding Azure Web PubSub client connection id.
+ * A `WebPubSubTranslator` instance is created for each Engine.IO server instance. It's designed to:
+ * 1. Manages all Azure Web PubSub client connections and keep them consistent with corresponding EIO clients.
+ * 2. Handle upstream invoke requests from AWPS and then translate them into Engine.IO behaviours.
+ * 3. Translates Engine.IO behaviours to AWPS behaviours like REST API calls.
+ * 4. Makes the EIO `sid` same as its corresponding Azure Web PubSub client connection id.
  */
-class WebPubSubTranslator {
-}
+export class WebPubSubTranslator {
+  /**
+   * Each `WebPubSubTranslator` instance is bound to a Engine.IO server instance and vice versa.
+   */
+  public linkedEioServer: BaseServer;
+
+  /**
+   * Client for connecting to a Web PubSub hub
+   */
+  private serviceClient: WebPubSubServiceClient;
+
+  /**
+   * Map from the `connectionId` of each client to its corresponding logical `ClientConnectionContext`.
+   */
+  private clientConnections: Map<string, ClientConnectionContext> = new Map();
+
+  /**
+   * Handle upstream invoke requests from AWPS.
+   */
+  private webPubSubEventHandler: WebPubSubEventHandler;
+
+  /**
+   * Options for Azure Web PubSub service.
+   */
+  private webPubSubOptions: WebPubSubExtensionOptions;
+
+  /**
+   * In native Engine.IO, the `sid` of each EIO connnection is generated by server randomly.
+   * As for AWPS, it generates `ConnectionId` for each client.
+   * For each EIO connection, the extension enforces its `sid` in server side is same as the ConnectionId assigned by service.
+   * This array stores all `ConnectionId` which is generated by AWPS and is prepared to be assigned to EIO connection.
+   */
+  private candidateSids: Array<string> = new Array();
+
+  constructor(server: BaseServer, options: WebPubSubExtensionOptions) {
+    debug("constructor");
+
+    this.linkedEioServer = server;
+    this.webPubSubOptions = options;
+    this.serviceClient = new WebPubSubServiceClient(
+      this.webPubSubOptions.connectionString,
+      this.webPubSubOptions.hub,
+      {
+        allowInsecureConnection: true,
+      }
+    );
+
+    this.webPubSubEventHandler = new WebPubSubEventHandler(this.webPubSubOptions.hub, {
+      path: this.webPubSubOptions.path,
+      handleConnect: async (req, res) => {
+        var connectionId = req.context.connectionId;
+        debug(`onConnect, connectionId = ${connectionId}`);
+
+        var context = new ClientConnectionContext(this.serviceClient, connectionId);
+
+        context.on(EIO_SERVER_ACCEPT_CONNECTION, (openPacketPayload: string) => {
```

Comment:
Emitted in [here](https://github.com/xingsy97/webpubsub-socketio-extension/blob/master/src/EIO/components/web-pubsub-transport.ts#L104)